ipinfo/cmd_download.go

```diff
@@ -0,0 +1,227 @@
+
+var completionsDownload = &complete.Command{
+	Flags: map[string]complete.Predictor{
+		"-c":         predict.Nothing,
+		"--compress": predict.Nothing,
+		"-f":         predict.Nothing,
+		"--format":   predict.Nothing,
+		"-t":         predict.Nothing,
+		"--token":    predict.Nothing,
+		"-h":         predict.Nothing,
+		"--help":     predict.Nothing,
+	},
+}
+
+func printHelpDownload() {
+	fmt.Printf(
+		`Usage: %s download [<opts>] <database> [<output>]
+
+Description:
+    Download the free ipinfo databases.
+
+Examples:
+    # Download country database in csv format.
+    $ %[1]s download country -f csv > country.csv
+    $ %[1]s download country-asn country_asn.mmdb
+
+Databases:
+    asn            free ipinfo asn database.
+    country        free ipinfo country database.
+    country-asn    free ipinfo country-asn database.
+
+Options:
+  General:
+    --token <tok>, -t <tok>
+      use <tok> as API token.
+    --help, -h
+      show help.
+
+Outputs:
+    --compress, -c
+	save the file in compressed format.
+	default: false.
+    --format, -f <mmdb | json | csv>
+     output format of the database file.
+     default: mmdb.
+`, progBase)
+}
+
+func cmdDownload() error {
+	var fTok string
+	var fFmt string
+	var fZip bool
+	var fHelp bool
+
+	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
+	pflag.StringVarP(&fFmt, "format", "f", "mmdb", "the output format to use.")
+	pflag.BoolVarP(&fZip, "compress", "c", false, "compressed output.")
+	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.Parse()
+
+	args := pflag.Args()[1:]
+	if fHelp || len(args) > 2 || len(args) < 1 {
+		printHelpDownload()
+		return nil
+	}
+
+	var token string
+	if fTok == "" {
+		token = gConfig.Token
+	}
+
+	// require token for download.
+	if token == "" {
+		return errors.New("downloading requires a token; login via `ipinfo login` or pass the `--token` argument")
+	}
+
+	// get download format and extension.
+	var format string
+	var fileExtension string
+	switch strings.ToLower(fFmt) {
+	case "mmdb":
+		format = "mmdb"
+		fileExtension = "mmdb"
+	case "csv":
+		format = "csv.gz"
+		fileExtension = "csv"
+	case "json":
+		format = "json.gz"
+		fileExtension = "json"
+	default:
+		return errors.New("unknown download format")
+	}
+
+	if fZip {
+		fileExtension = fmt.Sprintf("%s.%s", fileExtension, "gz")
+	}
+
+	// download the db.
+	switch strings.ToLower(args[0]) {
+	case "asn":
+		err := downloadDb("asn", format, token, fileExtension, fZip)
+		if err != nil {
+			return err
+		}
+	case "country":
+		err := downloadDb("country", format, token, fileExtension, fZip)
+		if err != nil {
+			return err
+		}
+	case "country-asn":
+		err := downloadDb("country_asn", format, token, fileExtension, fZip)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("database '%v' is invalid", args[0])
+	}
+
+	return nil
+}
+
+func downloadDb(name, format, token, fileExtension string, zip bool) error {
```

Comment:
Let's not use golangs feature here for multi-names under a single type, we avoid that IIRC in this codebase, makes it hard to read at a glance.